### ISSUE TITLE
[ES-MAML] 

### ISSUE THREAD

**Issue by TigerStone93** [google-research/google-research#499]
Dear @xingyousong 

I tried to run your ES-MAML code, but I failed.

In declaration phase, `blackbox_maml_objects.py` does import like below.

```
from es_maml.first_order import first_order_pb2
from es_maml.first_order import first_order_pb2_grpc

from es_maml.zero_order import zero_order_pb2
from es_maml.zero_order import zero_order_pb2_grpc
```

However, there is no `first_order_pbc2`, `first_order_pb2_grpc`, `zero_order_pb2` and `zero_order_pb2_grpc` in `es_maml/first_order` and `es_maml/zero_order` directory.

As result, an error occurred like below.
`ImportError: cannot import name 'first_order_pb2' from 'es_maml.first_order' (unknown location)`

Could you upload those things to provide workable ES_MAML code?

If I misunderstood the way to run your code, please give me some advise.

Thank you in advance.

**Comment by xingyousong** [comment]
Hi there, thanks for your question. Sorry about that - I didn't realize that the generated pb2, py_pb2, and py_pb2_grpc files wouldn't automatically get sent to this public repository. In Google's internal repository, I have a Bazel defining how the system should build extra libraries from the first_order.proto:
```
proto_library(
    name = "first_order_proto",
    srcs = ["first_order.proto"],
    has_services = 1,
    cc_api_version = 2,
)

py_proto_library(
    name = "first_order_py_pb2",
    has_services = 1,
    api_version = 2,
    deps = [":first_order_proto"],
)

google3_py_grpc_library(
    name = "first_order_py_pb2_grpc",
    srcs = [":first_order_proto"],
    deps = [":first_order_py_pb2"],
)
```

To use the exact same setup with protos I have, you should generate the libraries using GRPC library, e.g. see [an example here](https://github.com/tensorflow/serving/issues/42).

However, these proto definitions are only really useful if you're using GRPC as a way of communication - If you're using a different API to allow the client to communicate with the servers, then you probably don't need any of this and maybe can modify the code to directly use e.g. dictionaries instead of protos as messages.

Hope that helps!

**Comment by TigerStone93** [comment]
Thank you for the reply, @xingyousong! 

Maybe I understood your advise properly and finally generated pb2.py and pb2_grpc.py

[The link](https://github.com/tensorflow/serving/issues/42) that you give provides useful information but it is little bit out-of-date.

So the command below doesn't work.
`$ protoc -I=./ --python_out=./ --grpc_out=./ --plugin=protoc-gen-grpc=`which grpc_python_plugin` ./first_order.proto`

Alternatively, I refered [this link](https://stackoverflow.com/questions/34713861/python-grpc-protobuf-stubs-generation-issue-grpc-out-protoc-gen-grpc-plugin).

In conclusion, commands below work.
```
$ pip install protobuf
$ pip install grpcio-tools==1.32
$ pip install googleapis-common-protos

$ python -m grpc_tools.protoc --proto_path=. --python_out=. --grpc_python_out=. first_order.proto
```

**Comment by xingyousong** [comment]
Thanks for the commands! I've added this to the README: https://github.com/google-research/google-research/tree/master/es_maml

**Comment by TigerStone93** [comment]
@xingyousong 

Sorry for the interrupt again, but I got 2 more questions to ask...

[1]
`es_maml_server.py` and `es_maml_client.py` use a flag for `urdf_data_path`

```
def main(unused_argv):
  base_config = config_util.get_config(urdf_root=FLAGS.urdf_data_path)
  config = config_util.generate_config(
      base_config, current_time_string=FLAGS.current_time_string)
```

However, the flag is not declared like below.

```
flags.DEFINE_integer("server_id", 0, "The id of the server.")
flags.DEFINE_integer("port", 20000, "Port number.")
flags.DEFINE_string("current_time_string", "NA", "current time string")
```

As a result, an AttributeError is raised.

How could I solve it? Is it OK to eliminate `urdf_root` parameter?

[2]
And there is a dependency issue too.

First, grpc doesn't have loas2.

Maybe the configuration of grpc has been changed.

Vikram Agarwal said same issue on [this link](https://gitter.im/tensor2tensor/Lobby?at=5d5c6a736ddc404fbce2b346)

Anyway import below doesn't work at all.

```
import grpc
from grpc import loas2 #ImportError
```

Also, `grpc.server()` function doesn't have ports argument no more.

```
  server = grpc.server(
      futures.ThreadPoolExecutor(max_workers=100), ports=(port,)) #TypeError
```

Could you give any information of requisite library that you use?

Then It would help me and other users a lot.

In parallel, I will try to find and share any solution for this.

Thank you.

**Comment by xingyousong** [comment]
For the `urdf_root=FLAGS.urdf_data_path`, you can eliminate that FLAGS; sorry it was used internally for robotics applications and I forgot to delete. I'll update with a followup push later on.

For the `loas2` and `ports=...` parts, I'm importing the `grpc` from an internal library, but this library has been pushed to open-source here: https://github.com/grpc/grpc. I think some of the functions though like `loas2` are google specific (which simply returns my Google user-name). For `ports...`, it tells the locations that the aggregator client should connect to. I'm not too sure what to do with it when using the repository outside of Google.

Anyways, these implementations are Google-specific networking pipelines; if there's a better method to do client-server connecting on e.g. Cloud instances, you should use that framework instead.
